README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# measoshapes

The goal of measoshapes is to provide regionalization boundaries for
MEASO.

There are several in-built data sets currently at **version 05**. See
below where each is explained.

## Installation

To install measoshapes from Github use

``` r
## install.packages("remotes")
remotes::install_github("AustralianAntarcticDivision/measoshapes")
```

## Map

To map the regions with base plotting (with the [sf
package](https://r-spatial.github.io/sf)).

First join the names and properties to the geometry, we also group the
features so that the seam at the anti-meridian is removed.

``` r
library(measoshapes)
library(dplyr)
library(sf)

measo <- measo_regions05 %>% group_by(name) %>% summarize() %>% 
  inner_join(measo_names)
#> Joining, by = "name"
```

Now, plot the geometry with the offical colours and make sure no ‘reset’
so that we can add other layers.

``` r
plot(st_geometry(measo), col = measo$fill, reset = FALSE, border = NA)
```

<img src="man/figures/README-plot-1.png" width="100%" />

A very simple coastline data is added.

``` r
coast <- rnaturalearth::ne_countries(scale = "medium", returnclass = "sf") %>% dplyr::filter(sovereignt == "Antarctica")
coast <- st_transform(coast, 
                      st_crs(measo))

plot(st_geometry(measo), col = measo$fill, reset = FALSE, border = NA)
plot(st_geometry(coast), col = "#808080", add = TRUE, border = NA)
```

<img src="man/figures/README-coast-1.png" width="100%" />

## Map with ggplot2

A ggplot2 example. To use the literal fill colours we need
`scale_fill_identity()`, this is ggplot’s “straight-through” mechanism.

``` r

library(ggplot2)
ggplot(measo, aes(fill = fill)) + geom_sf(colour = NA) + scale_fill_identity() + geom_sf(data = coast, aes(fill = NULL))
```

<img src="man/figures/README-unnamed-chunk-1-1.png" width="100%" />

## Map with SOmap

The [SOmap
package](https://github.com/AustralianAntarcticDivision/SOmap) provides
some simpler ways of creating and adding to polar maps. Once a map is
set up it knows what projection it is in, so we can forget about having
to transform the data.

We can add arbitrary data at longitude/latitudes pairs and SOplot knows
what to do.

``` r
## atm we need a special branch (2020-02-20)
#remotes::install_github("AustralianAntarcticDivision/SOmap@dev-0.6")
library(SOmap)
#> Loading required package: raster
#> Loading required package: sp
#> 
#> Attaching package: 'raster'
#> The following object is masked from 'package:dplyr':
#> 
#>     select
SOmap(trim = -32) ## set up a polar map
## set overlay to half-transparent
SOplot(st_geometry(measo), border = NA, col = scales::alpha(measo$fill, 0.75))  ## no need to worry about the projection
#> Warning in if (class(x) == "character") {: the condition has length > 1 and only
#> the first element will be used
#> Warning in if (class(x) != "CRS") {: the condition has length > 1 and only the
#> first element will be used
#> Warning in SOproj(x = x, y = y, target = target, source = source, ...): assuming
#> generic data is in longitude,latitude

ll <- cbind(lon = c(147, 100, -10, -80), 
            lat = c(-42, -50, -70, -50))
SOplot(ll, pch = 1:4, col = viridis::viridis(4), cex = 3, lwd = 4)
#> Warning in SOproj(x = x, y = y, target = target, source = source, ...): assuming
#> generic data is in longitude,latitude

## other objects work too (sf, raster, sp)
SOplot(SOmap_data$fronts_park)
#> Warning in plot.sf(SObj, add = add, ...): ignoring all but the first attribute
SOplot(coast, col = "#808080")
#> Warning in plot.sf(SObj, add = add, ...): ignoring all but the first attribute
```

<img src="man/figures/README-SOmap-1.png" width="100%" />

## Data sets

There is no code in the measoshapes package, but the package records the
creation of each shapes layer using standard R mechanisms.

There are five datasets, and for normal usage, calculating areas, and
making maps we would use the first three. The `_coastline` forms have
the continent of Antarctica cut out of the shapes.

  - `measonames` a dataframe of the `name`, `sector`, and `zone`
  - `measo_regions05_ll_coastline` the polygons with the continent of
    Antarctica cut out
  - `measo_regions05_coastline` the polygons (in polar) with the
    continent of Antarctica cut out

Two other forms in polar and longitude/latitude form are use for model
overlays (for technical reasons).

  - `measo_regions05` the polygons of each combination of sector and
    zone
  - `measo_regions05_ll` the polygons in longitude/latitude form

To make a map with these we can use the following code

``` r
library(measoshapes)
plot(measo_regions05_ll_coastline, reset = FALSE)
maps::map(add = TRUE, col = "grey", fill = FALSE)
```

<img src="man/figures/README-measo-map-1.png" width="100%" />

To export to shapefile:

``` r
st_write(measo_regions05_ll_coastline, "myfile.shp", driver = "SHP")
```

Here we explore the names in more detail. The northern regions are
assigned codes that include “T” for **temperate**, but they aren’t
supposed to be part of MEASO. They share a sector but have no assigned
zone. From `colour_values` on zone they get black.

``` r
par(mar = rep(0.2, 4))
plot(st_geometry(measo_regions05_ll), reset = FALSE, 
     col = colourvalues::colour_values(measo_names$sector, alpha = 0.5))
cds <- st_coordinates(st_centroid(measo_regions05_ll))
#> Warning in st_centroid.sfc(st_geometry(x), of_largest_polygon =
#> of_largest_polygon): st_centroid does not give correct centroids for longitude/
#> latitude data
text(cds, lab = measo_names$name, cex = 0.7)
par(xpd = NA)
text(0, -20, "name")
maps::map(add = TRUE)
```

<img src="man/figures/README-dummy-1.png" width="100%" />

``` r
par(xpd = TRUE)

plot(st_geometry(measo_regions05_ll), reset = FALSE, 
     col = colourvalues::colour_values(measo_names$zone, alpha = 0.5))
cds <- st_coordinates(st_centroid(measo_regions05_ll))
#> Warning in st_centroid.sfc(st_geometry(x), of_largest_polygon =
#> of_largest_polygon): st_centroid does not give correct centroids for longitude/
#> latitude data
text(cds, lab = measo_names$zone, cex = 0.7)
par(xpd = NA)
text(0, -20, "zone")
maps::map(add = TRUE)
```

<img src="man/figures/README-dummy-2.png" width="100%" />

``` r
par(xpd = TRUE)


plot(st_geometry(measo_regions05_ll), reset = FALSE, 
     col = colourvalues::colour_values(measo_names$zone, alpha = 0.5))
cds <- st_coordinates(st_centroid(measo_regions05_ll))
#> Warning in st_centroid.sfc(st_geometry(x), of_largest_polygon =
#> of_largest_polygon): st_centroid does not give correct centroids for longitude/
#> latitude data
text(cds, lab = measo_names$sector, cex = 0.7)
par(xpd = NA)
text(0, -20, "sector")
maps::map(add = TRUE)
```

<img src="man/figures/README-dummy-3.png" width="100%" />

``` r
par(xpd = TRUE)
```

-----

Please note that the ‘measoshapes’ project is released with a
[Contributor Code of Conduct](CODE_OF_CONDUCT.md). By contributing to
this project, you agree to abide by its terms.


NEWS.md
# measoshapes 0.0.05.1

* Extend the shapes southerer. 

* Added colours. 

# measoshapes 0.0.05

* Release "05" creates "_coastline" versions in polar and in longitude/latitude. 
 There are now dummy polygons in the north for each sector, for classifying data that 
 is outside the primary regions. 


# measoshapes 0.0.04

* Release "04", renames "WestAtlantic" to "Atlantic"


data-raw/munge-sectors03.R
minlat <- -80
sectors_ll <- tibble::tribble(~lon, ~lat, ~zone,
                 -125, minlat,  1,
                 -125, -35,  1,

                 -70, minlat, 2,
                 -64.5, -66, 2,
                 -60, -64, 2,
                 -55.054477039271191, -63.260598133971456, 2,
                 -63.799776827053591, -54.721070105901589, 2,

               30, minlat, 3,
               30, -35, 3,

                 115, minlat, 4,
                 115, -35, 4,

                 158, minlat, 5,
                 158, -75,    5,
                 170, -71.60, 5,
                 170, -47.50, 5
                 )

library(spbabel)
library(dplyr)
library(sf)
sectors <- sectors_ll %>%
  transmute(lon, lat, x_ = lon, y_ = lat, branch_ = zone, object_ = zone, order_ = row_number()) %>%
  sp(crs = "+init=epsg:4326")
sectors <- sf::st_as_sf(sectors)
#file.copy("../measo-access/shapes/zones.Rdata", "data-raw/zones.Rdata")
load("data-raw/zones.Rdata")

#plot(sectors)
#plot(zones, add = TRUE)
#maps::map(add = TRUE)
domain <- st_cast(spex::polygonize(raster::raster(raster::extent(-180, 180, minlat, -30), nrows = 1,
                                                  ncols = 1,
                                                  crs = st_crs(sectors)$proj4string)), "LINESTRING")

lns <- st_sf(geometry = c(st_geometry(sectors),
                          st_geometry(domain),
                          st_geometry(zones)), crs = 4326)

measo_regions03g <- st_cast(st_polygonize(st_union(lns)))
## drop degenerate regions
measo_regions03g <- measo_regions03g[st_area(st_set_crs(measo_regions03g, NA)) > 1]


measo_regions03 <- st_sf(geometry = measo_regions03g)
#st_coordinates(st_centroid(measo_regions03))[,2]
plot(measo_regions03[st_coordinates(st_centroid(measo_regions03))[,2] < -38, ],
     col = sample(rainbow(nrow(measo_regions03)-1)))




## order by longitude, then latitude of bottom left corner
ord <- spbabel::sptable(measo_regions03) %>%
  group_by(object_) %>%
  arrange(x_, y_) %>% slice(1) %>% ungroup() %>% arrange(x_, y_) %>% pull(object_)

measo_regions03 <- measo_regions03[ord, ]

measo_regions03$name <- c("WPA", "WPS", "WPN",
                          NA, ## northern background,
                          "EPA", "EPS", "EPN",
                          "WAA", "WAS", "WAN",

                          "CIA", "CIS", "CIN",
                          "EIA", "EIS", "EIN",
                          "WPA", "WPS", "WPN")
measo_regions03$a <- NULL
measo_regions03_ll <- measo_regions03
plot(st_geometry(measo_regions03), reset = FALSE,
     col = rainbow(length(unique(measo_regions03$name)), alpha = 0.4)[factor(measo_regions03$name)], border  = NA)
text(st_coordinates(st_centroid(measo_regions03)),
     lab = measo_regions03$name)
sp::plot(orsifronts::orsifronts, add = TRUE)


## zones polar
measo_regions03 <- sf::st_transform(sf::st_set_crs(sf::st_segmentize(sf::st_set_crs(measo_regions03_ll, NA), 0.2), 4326),
                          "+proj=laea +lat_0=-90 +lon_0=0 +datum=WGS84")


zz <- c("Antarctic", "Subantarctic", "Northern")
sec <- c("WestPacific", "EastPacific", "WestAtlantic",
         "CentralIndian", "EastIndian", "WestPacific")
measo_names <- tibble::tibble(name = measo_regions03$name,
                              sector = c(rep(sec[1], 3), NA,
                                         rep(sec[-1], each =  3)),
                              zone = c(zz,
                                       NA,
                                       rep(zz, 5)))
usethis::use_data(measo_names)
usethis::use_data(measo_regions03_ll)
usethis::use_data(measo_regions03)



data-raw/munge-sectors05.R
minlat <- -85
maxlat <- -30
library(sp)
library(reproj)
sectors_ll <- tibble::tribble(~lon, ~lat, ~zone,
                 -125, minlat,  1,
                 -125, maxlat,  1,

                 -70, minlat, 2,
                 -64.5, -66, 2,
                 -60, -64, 2,
                 -55.054477039271191, -63.260598133971456, 2,
                 -63.799776827053591, -54.721070105901589, 2,
                 -63.799776827053591, maxlat,2,

               30, minlat, 3,
               30, maxlat, 3,

                 115, minlat, 4,
                 115, maxlat, 4,

                 158, minlat, 5,
                 158, -75,    5,
                 170, -71.60, 5,
                 170, -47.50, 5,
                 170, maxlat, 5,
                 )

library(spbabel)
library(dplyr)
library(sf)
sectors <- sectors_ll %>%
  transmute(lon, lat, x_ = lon, y_ = lat, branch_ = zone, object_ = zone, order_ = row_number()) %>%
  sp(crs = "+init=epsg:4326")
sectors <- sf::st_as_sf(sectors)
#file.copy("../measo-access/shapes/zones.Rdata", "data-raw/zones.Rdata")
load("data-raw/zones.Rdata")
zones <- rbind(zones,
    st_sf(ID = 0, front = NA_character_,
          geometry = st_sfc(st_linestring(cbind(c(-180, 180), maxlat)))))
#plot(sectors)
#plot(zones, add = TRUE)
#maps::map(add = TRUE)
domain <- st_cast(spex::polygonize(raster::raster(raster::extent(-180, 180, minlat, -30), nrows = 1,
                                                  ncols = 1,
                                                  crs = st_crs(sectors)$proj4string)), "LINESTRING")

lns <- st_sf(geometry = c(st_geometry(sectors),
                          st_geometry(domain),
                          st_geometry(zones)), crs = 4326)

gg <- st_cast(st_polygonize(st_union(lns)))
## drop degenerate regions
gg <- gg[st_area(st_set_crs(gg, NA)) > 1]


measo_regions05 <- st_sf(geometry = gg)
plot(measo_regions05, col = sample(viridis::viridis(30)))



## order by longitude, then latitude of bottom left corner
ord <- spbabel::sptable(measo_regions05) %>%
  group_by(object_) %>%
  arrange(x_, y_) %>% slice(1) %>% ungroup() %>% arrange(x_, y_) %>% pull(object_)

measo_regions05 <- measo_regions05[ord, ]
plot(measo_regions05, col = viridis::viridis(25))
measo_regions05$name <- c("WPA", "WPS", "WPN",
                          "WPT", ## northern background,
                          "EPA", "EPS", "EPN",
                          "EPT",
                          "AOA", "AOT", "AOS",
                          "AON",
                          "CIA", "CIS", "CIN",
                          "CIT",
                          "EIA", "EIS", "EIN",
                          "EIT",
                          "WPA", "WPS", "WPN",
                          "WPT")
measo_regions05$a <- NULL
measo_regions05_ll <- measo_regions05
plot(st_geometry(measo_regions05), reset = FALSE,
     col = rainbow(length(unique(measo_regions05$name)),
                   alpha = 0.4)[factor(measo_regions05$name)], border  = NA)
text(st_coordinates(st_centroid(measo_regions05)),
lab = measo_regions05$name)
#   lab = 1:24)




sp::plot(orsifronts::orsifronts, add = TRUE)


## zones polar
measo_regions05 <- sf::st_transform(sf::st_set_crs(sf::st_segmentize(sf::st_set_crs(measo_regions05_ll, NA), 0.2), 4326),
                          "+proj=laea +lat_0=-90 +lon_0=0 +datum=WGS84")


zz <- c("Antarctic", "Subantarctic", "Northern", NA,
        "Antarctic", "Subantarctic", "Northern", NA,
        "Antarctic", NA, "Subantarctic", "Northern",
        "Antarctic", "Subantarctic", "Northern", NA,
        "Antarctic", "Subantarctic", "Northern", NA,
        "Antarctic", "Subantarctic", "Northern", NA)
sec <- c("WestPacific", "EastPacific", "Atlantic",
         "CentralIndian", "EastIndian", "WestPacific")
fill <- c("#04405CFF",
          "#054e70FF",
          "#016074FF",
          "#1094AFFF",
          "#00AFD5FF",
          "#BCECFEFF",
          "#4B7D7EFF",
          "#5F9EA0FF",
          "#EAFAFFFF",
          "#FFFFFFFF",
          "#000000FF",
          "#52575AFF")[c(3,4,2,NA,
                         10,8,7,NA,
                         6,NA,4,2,
                         10,8,7,NA,
                         9,5,3,NA,
                         6,4,2,NA)]
fill[is.na(fill)] <- NA
measo_names <- tibble::tibble(name = measo_regions05$name,
                              sector = rep(sec, each = 4),
                              zone = zz,
                              fill = fill)
usethis::use_data(measo_names, overwrite = TRUE)
usethis::use_data(measo_regions05_ll, overwrite = TRUE)
usethis::use_data(measo_regions05, overwrite = TRUE)

l <- raadtools:::keepOnlyMostComplexLine(sp::SpatialLinesDataFrame(as(sf::as_Spatial(SOmap::SOmap_data$ant_coast_land), "SpatialLines"), data.frame(a=1), match.ID = F))
mm <- coordinates(l)[[1]][[1]]

mm <- reproj(mm, source = "+proj=stere +lat_0=-90 +lat_ts=-71 +lon_0=0 +k=1 +x_0=0 +y_0=0 +datum=WGS84 +units=m +no_defs +ellps=WGS84 +towgs84=0,0,0", target = 4326)[, 1:2]
aa <- st_union(st_cast(sfdct::ct_triangulate(st_sfc(st_polygon(list(mm))))))
aa <- st_set_crs(aa, st_crs(measo_regions05_ll))
measo_regions05_ll_coastline <- st_cast(st_difference(measo_regions05_ll, aa))
usethis::use_data(measo_regions05_ll_coastline, overwrite = TRUE)
measo_regions05_coastline <- sf::st_transform(sf::st_set_crs(sf::st_segmentize(sf::st_set_crs(measo_regions05_ll_coastline, NA), 0.2), 4326),
                                    "+proj=laea +lat_0=-90 +lon_0=0 +datum=WGS84")

usethis::use_data(measo_regions05_coastline, overwrite = TRUE)



data-raw/munge-sectors04.R
minlat <- -80
sectors_ll <- tibble::tribble(~lon, ~lat, ~zone,
                 -125, minlat,  1,
                 -125, -35,  1,

                 -70, minlat, 2,
                 -64.5, -66, 2,
                 -60, -64, 2,
                 -55.054477039271191, -63.260598133971456, 2,
                 -63.799776827053591, -54.721070105901589, 2,

               30, minlat, 3,
               30, -35, 3,

                 115, minlat, 4,
                 115, -35, 4,

                 158, minlat, 5,
                 158, -75,    5,
                 170, -71.60, 5,
                 170, -47.50, 5
                 )

library(spbabel)
library(dplyr)
library(sf)
sectors <- sectors_ll %>%
  transmute(lon, lat, x_ = lon, y_ = lat, branch_ = zone, object_ = zone, order_ = row_number()) %>%
  sp(crs = "+init=epsg:4326")
sectors <- sf::st_as_sf(sectors)
#file.copy("../measo-access/shapes/zones.Rdata", "data-raw/zones.Rdata")
load("data-raw/zones.Rdata")

#plot(sectors)
#plot(zones, add = TRUE)
#maps::map(add = TRUE)
domain <- st_cast(spex::polygonize(raster::raster(raster::extent(-180, 180, minlat, -30), nrows = 1,
                                                  ncols = 1,
                                                  crs = st_crs(sectors)$proj4string)), "LINESTRING")

lns <- st_sf(geometry = c(st_geometry(sectors),
                          st_geometry(domain),
                          st_geometry(zones)), crs = 4326)

measo_regions03g <- st_cast(st_polygonize(st_union(lns)))
## drop degenerate regions
measo_regions03g <- measo_regions03g[st_area(st_set_crs(measo_regions03g, NA)) > 1]


measo_regions04 <- st_sf(geometry = measo_regions03g)
#st_coordinates(st_centroid(measo_regions03))[,2]
plot(measo_regions04[st_coordinates(st_centroid(measo_regions04))[,2] < -38, ],
     col = sample(rainbow(nrow(measo_regions04)-1)))




## order by longitude, then latitude of bottom left corner
ord <- spbabel::sptable(measo_regions04) %>%
  group_by(object_) %>%
  arrange(x_, y_) %>% slice(1) %>% ungroup() %>% arrange(x_, y_) %>% pull(object_)

measo_regions04 <- measo_regions04[ord, ]

measo_regions04$name <- c("WPA", "WPS", "WPN",
                          NA, ## northern background,
                          "EPA", "EPS", "EPN",
                          "AOA", "AOS", "AON",

                          "CIA", "CIS", "CIN",
                          "EIA", "EIS", "EIN",
                          "WPA", "WPS", "WPN")
measo_regions04$a <- NULL
measo_regions04_ll <- measo_regions04
plot(st_geometry(measo_regions04), reset = FALSE,
     col = rainbow(length(unique(measo_regions04$name)),
                   alpha = 0.4)[factor(measo_regions04$name)], border  = NA)
text(st_coordinates(st_centroid(measo_regions04)),
     lab = measo_regions04$name)
sp::plot(orsifronts::orsifronts, add = TRUE)


## zones polar
measo_regions04 <- sf::st_transform(sf::st_set_crs(sf::st_segmentize(sf::st_set_crs(measo_regions04_ll, NA), 0.2), 4326),
                          "+proj=laea +lat_0=-90 +lon_0=0 +datum=WGS84")


zz <- c("Antarctic", "Subantarctic", "Northern")
sec <- c("WestPacific", "EastPacific", "Atlantic",
         "CentralIndian", "EastIndian", "WestPacific")
measo_names <- tibble::tibble(name = measo_regions04$name,
                              sector = c(rep(sec[1], 3), NA,
                                         rep(sec[-1], each =  3)),
                              zone = c(zz,
                                       NA,
                                       rep(zz, 5)))
usethis::use_data(measo_names, overwrite = TRUE)
usethis::use_data(measo_regions04_ll)
usethis::use_data(measo_regions04)


